[spring-ecommerce-product - Step-2] Submit miooo0o's mission

miooo0o opened the pull request.

Wordbe commented on src/main/kotlin/ecommerce/service/LoginMemberArgumentResolver.kt:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.web.method.support.ModelAndViewContainer
+
+@Component
+class LoginMemberArgumentResolver(
```
Instead of placing it in the `service` package, try moving it to a package similar to `webconfig`. It might improve cohesion. 🙂

miooo0o replied:
I’ve applied the change and moved package to the config. I hope I understood it correctly!

Wordbe commented on src/main/kotlin/ecommerce/service/LoginMemberArgumentResolver.kt:
```diff
@@ -0,0 +1,39 @@
+            webRequest.getHeader("Authorization")
+                ?: throw UnauthorizedException("Authorization header is missing")
+
+        return authHeader.removePrefix("Bearer ").trim()
```
Try defining magic numbers as constants. 🔢

Wordbe replied:
It looks like the token might be extracted even if `Bearer` isn’t included. If `Bearer` is required, please add appropriate validation! 🙂

Wordbe commented on src/main/kotlin/ecommerce/repository/MemberRepository.kt:
```diff
@@ -0,0 +1,100 @@
+        val sql = MemberConstsSQL.COUNT_BY_EMAIL
+        val existing = jdbcTemplate.queryForObject(sql, Int::class.java, email)
+        return existing != null && existing > 0
+    }
```
Try using an `EXISTS` query instead of `COUNT`. Let’s also explore how they differ!

Wordbe commented on src/main/kotlin/ecommerce/controller/AdminController.kt:
```diff
@@ -0,0 +1,42 @@
+    fun getActiveMembers(): ResponseEntity<List<MemberStatsResponse>> {
+        val products = service.getActiveMembers(ACTIVE_MEMBERS_LAST_7_DAYS)
+        return ResponseEntity.ok(products)
+    }
```
If you’re not explicitly setting a status code, try removing `ResponseEntity`. Thanks to `@RestController` including `@ResponseBody`, it’ll still work! 🙂

Wordbe commented on src/main/kotlin/ecommerce/controller/CartController.kt:
```diff
@@ -0,0 +1,62 @@
+    ): ResponseEntity<CartResponse> {
+        val response = service.upsertCartItem(member.id, productId, request.quantity)
+        return ResponseEntity.status(HttpStatus.CREATED).body(response)
+    }
```
When naming resources in a REST API, it's better to use a single name like `cart-items` instead of separating with `/` as in `cart/items`. 🙂

Wordbe commented on src/main/kotlin/ecommerce/controller/ProductController.kt:
```diff
@@ -0,0 +1,76 @@
+        } catch (exception: Exception) {
+            ResponseEntity.status(HttpStatus.BAD_REQUEST).build()
+        }
+    }
```
This controller looks a bit too bulky. What about creating a Service to separate concerns, like in other parts of the code? 🙂

miooo0o replied:
Agree! Product was one of the first parts I wrote, so even to me, it feels quite messy.
I’ve applied the change in [this commit](https://github.com/herotechcourse/spring-ecommerce-product/pull/34/commits/beeaa2db83b80ce7c8179230713ee17be170c7c3)!

Wordbe replied:
Great Job. Thank you.

Wordbe commented on src/main/kotlin/ecommerce/helper/JdbcHelper.kt:
```diff
@@ -0,0 +1,21 @@
+import java.sql.PreparedStatement
+
+object JdbcHelper {
+    fun <Res> insertAndReturnKey(
```
That looks really convenient! 😄

Wordbe commented on src/main/kotlin/ecommerce/repository/CartRepository.kt:
```diff
@@ -0,0 +1,74 @@
+        val sql = CartConstsSQL.COUNT_BY_MEMBER
+        val existing = jdbcTemplate.queryForObject(sql, Int::class.java, memberId)
+        return existing != null && existing > 0
+    }
```
Consider using an `EXISTS` query here as well. 🙂

Wordbe commented on src/main/kotlin/ecommerce/repository/CartRepository.kt:
```diff
@@ -0,0 +1,74 @@
+
+    fun insert(
+        memberId: Long,
+        request: CartRequest,
```
It would be better to prevent DTOs from reaching the Repository layer. 🙂

Wordbe commented on src/main/kotlin/ecommerce/repository/MemberRepository.kt:
```diff
@@ -0,0 +1,100 @@
+    private fun prepareInsertStatement(
+        connection: Connection,
+        sql: String,
+        request: MemberRequest,
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/repository/MemberRepository.kt:
```diff
@@ -0,0 +1,100 @@
+        expected: String,
+    ): Boolean {
+        return actual == expected
+    }
```
It might be better to move this elsewhere—to a place that fits better! 🙂

Wordbe commented on src/main/kotlin/ecommerce/repository/ProductRepository.kt:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    fun existsById(id: Long): Boolean {
+        val sql = ProductConstsSQL.COUNT_BY_ID
```
Same here.

Wordbe commented on src/main/kotlin/ecommerce/repository/ProductRepository.kt:
```diff
@@ -0,0 +1,102 @@
+        return found != null && found > 0
+    }
+
+    fun insertWithKeyholder(request: ProductRequest): Long {
```
Same here!

Wordbe commented on src/main/kotlin/ecommerce/sql/StateConstsSQL.kt:
```diff
@@ -0,0 +1,40 @@
+        ORDER BY %s
+        """
+
+    const val ACTIVE_MEMBERS_TEMPLATE = """
```
This is not being used! 🚫

Wordbe commented on src/main/kotlin/ecommerce/service/AuthService.kt:
```diff
@@ -0,0 +1,30 @@
+@Service
+class AuthService(
+    private val repository: MemberRepository,
+    private val jwtTokenProvider: JwtTokenProvider,
```
`JwtToken` stands for JSON Web Token Token. You can remove the redundant word to make it clearer! 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/AuthService.kt:
```diff
@@ -0,0 +1,30 @@
+    private fun getPayloadFromToken(token: String): Long {
+        return jwtTokenProvider.getPayload(token).toLongOrNull()
+            ?: throw InvalidTokenException("can not parse token")
+    }
```
Personally, since it’s simple, I don’t think it needs to be extracted into a separate method! 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/CartService.kt:
```diff
@@ -0,0 +1,52 @@
+        return repository.selectByMemberId(memberId)
+    }
+
+    fun upsertCartItems(
```
The method name is `upsert`, but it seems to only perform an insert. 🙂

miooo0o replied:
I mistakenly thought I was updating it! I’ve reflected the change in [this commit](https://github.com/herotechcourse/spring-ecommerce-product/pull/34/commits/d8122cfe5d7185272a785dc5e56dc7b5e61c9517)

Wordbe commented on src/main/kotlin/ecommerce/service/JwtTokenProvider.kt:
```diff
@@ -0,0 +1,64 @@
+    private lateinit var rawSecretKey: String
+
+    @Value("\${security.jwt.token.expire-length}")
+    private var validityInMilliseconds: Long = 0
```
There are various ways to load variables in Spring—try the method below as well and compare! 🙂
```
@ConfigurationProperties(prefix = "security.jwt.token")
class JwtProperties(
    val secretKey: String,
    val expireLength: Long
)
```

Wordbe replied:
You don’t have to apply the change, but could you share the reason why? 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/JwtTokenProvider.kt:
```diff
@@ -0,0 +1,64 @@
+import javax.crypto.SecretKey
+
+@Component
+class JwtTokenProvider {
```
Since JWT is a critical part of the logic, consider adding test cases for it. ☺️

Wordbe commented on src/main/kotlin/ecommerce/sql/SortOption.kt:
```diff
@@ -0,0 +1,13 @@
+package ecommerce.sql
+
+enum class SortOption(val sql: String) {
+    PRODUCT_COUNT("add_count DESC"),
```
Update the hardcoded parts to use this enum instead! 🙂

Wordbe commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,32 @@
+security:
+  jwt:
+    token:
+      secret-key: TGZtZ3VyaU94UVZ1d2k2NGhNVmxuY3AzZ1JHVWltcE5HTnZkNWQzZg==
+      expire-length: 3600000
```
Shall we think about a proper way to handle sensitive information? 😀

Wordbe commented on src/test/kotlin/ecommerce/service/CartServiceTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+@SpringBootTest
+@Transactional
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
This annotation can cause significant performance degradation — it might be a good idea to look for an alternative. 🧐

miooo0o replied:
Thanks for letting me know! I’ve updated it accordingly.
Most of test changed happen in [here](https://github.com/herotechcourse/spring-ecommerce-product/pull/34/commits/1906614ea9cb0fcffd86f90162785537b1bade0d)

Wordbe replied:
Good. Thank you

Wordbe commented on src/test/kotlin/ecommerce/service/CartServiceTest.kt:
```diff
@@ -0,0 +1,49 @@
+@SpringBootTest
+@Transactional
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@Rollback
```
What about removing unnecessary annotations? (Try to find using this with `@Transactional`)

miooo0o replied:
Now I know now that both `@Transactional` and `@Rollback` are already included in `@SpringBootTest` 🎊

Wordbe replied:
No, `@SpringBootTest` itself does not include `@Transactional` or `@Rollback`.
Please double-check that! 🧐

Wordbe commented on src/test/kotlin/ecommerce/repository/CartRepositoryTest.kt:
```diff
@@ -0,0 +1,55 @@
+@SpringBootTest
+@Transactional
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.ANY)
+@ActiveProfiles("test")
```
Did you create a separate profile? 🧐

miooo0o replied:
That setup was honestly nonsense. I just threw it together when I didn’t know better.
I realized too late that the tests had serious issues.
I fixed a lot of it last week and will properly review and choose annotations next time.
Looking back now, it’s honestly quite embarrassing.

Wordbe commented on src/test/kotlin/ecommerce/repository/CartRepositoryTest.kt:
```diff
@@ -0,0 +1,55 @@
+
+@SpringBootTest
+@Transactional
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.ANY)
```
Why did you use this? 🤔

miooo0o replied:
I have to admit, I just combined annotations until the tests ran.
They weren’t well structured tests 🥲 just something I made work at the time.
I’ve realized this was a poor approach, but honestly, at that moment, it felt like the only way to make it.

Wordbe replied:
You've switched to using `@DataJdbcTest`. What does it do? 🙂

Wordbe commented on src/main/kotlin/ecommerce/service/LoginMemberArgumentResolver.kt:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.web.method.support.ModelAndViewContainer
+
+@Component
+class LoginMemberArgumentResolver(
```
I noticed you didn’t implement an `Interceptor`.
Using an `ArgumentResolver` seems sufficient, but I’m curious—was there a particular reason for that choice? 🙂

miooo0o replied:
I wasn’t ready to use the concept of an interceptor.
I wanted to finish it as quickly as I could rather than slowing down by introducing more a new concept I wasn’t familiar with. I believe that was the right decision. I’m curious to hear your thoughts.

Wordbe replied:
Yes, if you have time, try applying an interceptor as well. There’s a lot you can learn from it! 🚀

Wordbe reviewed (changes requested):
Hello Mina!

I’ve reviewed the code you submitted.
I can see that you put a lot of care into it. Your logic is well thought out and carefully implemented.
Great job!

Please check the feedback I’ve provided, and apply the suggestions where possible.

Thank you! 😊

miooo0o commented:
### Step 2 Refactor

Hello. My PR has been delayed much longer than expected. I apologize for the delay. 
I tried my best to incorporate the reviews you suggested.

---

1. LoginMemberArgumentResolver
    - Moved to `webconfig` package
    - Added constants for magic strings
    - Enforced `"Bearer "` prefix validation
2. ProductController / Service
    - Extracted controller logic to service layer

3. Repository Layer
    - Replaced `COUNT` queries with `EXISTS` (`Member`, `Cart`, `Product`)
    - Prevented DTO usage in repositories
    - Moved `matches()` method from `MemberRepository` → `Member` entity

4. Admin & Cart Controllers
    - Removed redundant `ResponseEntity`
    - Renamed `/api/cart/items/{productId}` → `/api/cart-items/{productId}`

5. CartService
    - Renamed `upsertCartItems` → `insertCartItem`

6. JWT / Auth
    - Renamed `JwtTokenProvider` → `JwtProvider`
    - Used `@ConfigurationProperties` for JWT settings
    - Simplified token validation
7. SQL / Enum
    - Removed unused SQL constants
    - Replaced hardcoded sort strings with `SortOption` enum
8. Config & Secrets
    - Proper handling of `secret-key` in `application.yml`
        - But I missed that part while reviewing the test, put secret-key in `env` would be an idea?
9. Domain
   - Added an intermediate domain layer (e.g., `New**`) to handle entity creation.
     This approach is not fully polished, but I wanted to apply the feedback.
10. Tests
   **The tests have changed significantly** I am sorry!
   - Removed `@DirtiesContext` for better test performance and replace all tests to `mockMvc`
     - Unfortunately, I failed to fix the tests. When I removed `@DirtiesContext`, My bad tests seemed to be linked to
       each other, making it impossible to have properly isolated tests. So I decided to rewrite the tests using mocks.
   - Added JWT-related critical logic tests

Wordbe reviewed (approved):
Hello Mina!

I know I left quite a few review comments, didn’t I? 😊
Still, thank you for reading through each one and applying them carefully.

Through this PR, I believe you’ve gained solid experience working with JDBC in Spring, building well-structured APIs, and implementing authentication.
For any parts you couldn’t get to this time, I recommend revisiting and trying to implement them when you have time. It’ll be a great learning opportunity.

There are still a few comments that haven’t been marked as resolved, so please take a moment to check those.

Overall, you’ve reflected the feedback well and structured the project nicely, so I’ll go ahead and approve and merge this PR.

Great work so far, and best of luck on the next mission too!
Thank you! 🙌

miooo0o commented:
Hi, Suno!

- In response to the [comment](https://github.com/herotechcourse/spring-ecommerce-product/pull/34#discussion_r2256835999),
I wasn't sure how to handle tokens. Now I know that you can solve it using `.env`!

- In response to the [comment](https://github.com/herotechcourse/spring-ecommerce-product/pull/34#discussion_r2256841852),
Yes, you're right. I was confusing it with another annotation. `@SpringBootTest` is very slow and is suitable for integration testing because it loads all beans. As you mentioned, it doesn't include Transactional . `@DataJdbcTest` loads only Jbdc-related beans and includes Transactional . I heard it's a bit faster, so I applied it.

And thanks for the note about the interceptor. I'd like to come back and try it out later when I have time.
